libraries/chain/controller.cpp

```diff
@@ -520,6 +522,28 @@ struct controller_impl {
          const auto hash = calculate_integrity_hash();
          ilog( "database initialized with hash: ${hash}", ("hash", hash) );
       }
+
+      if (head && fork_db.pending_head() && fork_db.root()) {
+         if (read_mode != db_read_mode::IRREVERSIBLE) {
+            if (head->block_num < fork_db.pending_head()->block_num) { 
+               // irreversible mode => speculative mode
+               wlog("db_read_mode has been changed: forwarding state from block ${h} to fork pending head ${fh}", ("h", head->block_num)("fh", fork_db.pending_head()->block_num));
+               // let's go forward from lib to pending head, and set head as pending head
+               maybe_switch_forks( fork_db.pending_head(), controller::block_status::validated );
+            }
+         } else {
+            // speculative mode => irreversible mode
```

Comment:
Let's just simplify for now by disallowing switching from speculative mode to irreversible mode without a replay.

So if in irreversible mode, assert that `lib_num >= head->block_num`. But do this check just prior to the line:
```cpp
bool report_integrity_hash = !!snapshot || (lib_num > head->block_num);
```

Reply:
Alternatively, if you have already tested things and it still works fine with the proposed changes, I guess we can keep it. But I think it should be moved up to before the `if( report_integrity_hash ) {` line.